daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -203,31 +201,27 @@ abstract class SequenceParserBase(
               pstate.mpstate.moveOverOneGroupIndexOnly()
             }
 
-            val newLastChildNode = pstate.infoset.contents.lastOption
-            if (newLastChildNode != savedLastChildNode) {
-              // We have added at least one child to to this complex during
+            val newLastChildNode = pstate.infoset.maybeLastChild
+            if (newLastChildNode.isDefined) {
+              // We have potentially added a child to to this complex during
               // this array loop.
               //
               // If the new child is a DIArray, we know this DIArray has at
               // least one element, but we don't know if we actually added a
               // new one in this loop or not. So just get the last array
-              // element and set it as final anyways. Note that we need a null
-              // check because if we didn't add an array element this go
-              // around, the last element may have already been walked and
-              // freed and so could now be null.
+              // element and set it as final anyways.
               //
               // If it's not a DIArray, that means it's just an optional
               // simple/complex and that will get set final below where all
               // other non-array elements get set as final.
-              newLastChildNode.get match {
-                case a: DIArray => {
-                  val lastArrayElem = a.contents.last
-                  if (lastArrayElem != null) {
-                    lastArrayElem.setFinal()
-                    pstate.walker.walk()
-                  }
+              val lastChild = newLastChildNode.get
+              if (!lastChild.isSimple && !lastChild.isComplex) {
```

Comment:
Two calls to isSimple and isComplex? Let's just add an isArray method to go with the others. It's again disappointing that x.isInstanceOf[DIArray] isn't every bit as fast as having an isArray boolean. If isSimple and isComplex are virtual functions defined on DINode and DISimple and DIComplex, then I fail to understand why they wouldn't be pretty much identical in preformance to x.isInstanceOf, why would this be? 

This really suggests we should have an Int member which we initialize to an integer enum value that identifies simple/complex/array, and then we can have inline final def isSimple = nodeKindEnum eq SIMPLE and similar for isComplex and isArray. These would then not be virtual functions at all. But I don't like having to play that sort of tricks. We're just reinventing virtual dispatch, doing so in a less general way that we know will just turn into an integer equality test.

Reply:
Yeah, I'm surprised to. Of all the changes, I think this one didn't make a huge difference, it was definitely withing the margin of error for Java. But it did seem to consistent error on the positive side of that difference so I figured I'd keep it in to try to eek out every bit of performance. I'll add an isArray so this is at least more readable.